### PR TITLE
feat(syntax): allow comments in hashes and before EOF

### LINKED ISSUE #15561: Allow comments in pipeline config between hash entries
Currently it seems not allowed to make comments between hash entries, this is a feature request to allow it.

That means the following is valid:

```
mutate {
  rename =>  {
    # before entries
    "fieldA" => # inside entry
      "fieldB"
    "fieldC" => "fieldD"
    # after entries
  }
}
```

But that this config is invalid, which i think is unfortunate:

```
mutate {
  rename =>  {
    # before entries
    "fieldA" => "fieldB"
    # between entries (currently illegal)
    "fieldC" => "fieldD"
    # after entries
  }
}
```

The relevant part of the current config grammar (I'm not sure of the difference between the lscl_grammar and the "regular" one, but they both share the same issue):

```
  rule hash
    "{"
      cs
      hashentries?
      cs
    "}"
    <LogStash::Compiler::LSCL::AST::Hash>
  end

  rule hashentries
    hashentry (whitespace hashentry)*
    <LogStash::Compiler::LSCL::AST::HashEntries>
  end

  rule hashentry
    name:(number / bareword / string) cs "=>" cs value
    <LogStash::Compiler::LSCL::AST::HashEntry>
  end
```

Here `cs` is optional comment or whitespace, but `whitespace` is *only* whitespace. So comments are allowed before, after, and *inside* a hash entry, *but not between them*. I think that putting `cs` in there after `whitespace` would fix the issue.

### LINKED ISSUE #15651: Pipeline Grammar fails to parse comment that terminates on EOF
There is a corner case when that when a comment symbol is present on EOF on a Logstash Pipeline configuration, Logstash errors out. An example of this is when a pipeline is set such as:

```
input { .. }
output { .. }
# this is a comment<EOF>
```

And the error is something like:

```
[2023-11-24T16:21:35,982][ERROR][logstash.agent           ] Failed to execute action {:action=>LogStash::PipelineAction::Create/pipeline_id:main, :exception=>"LogStash::ConfigurationError", 
:message=>"Expected one of [^\\r\\n], \"\\r\", \"\\n\" at line 29, column 2 (byte 609) after #", :backtrace=>["/usr/share/logstash/logstash-core/lib/logstash/compiler.rb:32:in `compile_imperative'",
"org/logstash/execution/AbstractPipelineExt.java:239:in `initialize'", "org/logstash/execution/AbstractPipelineExt.java:173:in `initialize'", "/usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb:48:in `initialize'", 
"org/jruby/RubyClass.java:931:in `new'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline_action/create.rb:49:in `execute'", "/usr/share/logstash/logstash-core/lib/logstash/agent.rb:386:in `block in converge_state'"]}
```

**Workaround**
Ensure EOF is on a new line.

### PR DESCRIPTION
**Pull request by jonaslb**
## Release notes

Fixed the syntax rules to allow comments between hash entries and immediately before the end of a file. Comments in these places would previously give syntax errors.

## What does this PR do?

In the grammer definitions for hashes, `whitespace` was replaced with `cs` to allow either whitespace _or_ comments. 
Additionally, the grammar definition for comments was previously required to end with a newline, now it can end with a newline _or_ EOF, using the "not anything" treetop rule `!.`.

## Why is it important/What is the impact to the user?

Users get surprised when they find out that comments are not allowed in hashes or at the end of a file.
This is allowed in practically all other languages, including ruby.
This PR thus aligns Logstash syntax with user expectations.

## Checklist

- [x] My code follows the style guidelines of this project
    - AFAICT, no automated check for this? I believe it follows the guide though.
- [x] I have commented my code, particularly in hard-to-understand areas
    - I only added syntax, generated code (treetop), and test code. Aside from the generated code, changes are trivial.
- [x] I have made corresponding changes to the documentation
    - No docs changes should be needed.
- [x] I have made corresponding change to the default configuration files (and/or docker env variables)
    - Should also not be needed.
- [x] I have added tests that prove my fix is effective or that my feature works
    - Yes though!

## Author's Checklist

- [ ]

## How to test this PR locally

To run the test that I added, execute `SPEC_OPTS="-fd -P logstash-core/spec/logstash/config/config_ast_spec.rb" ./gradlew :logstash-core:rubyTests --tests org.logstash.RSpecTests` (this command is copied from `README.md` but with the path changed).

Alternatively, create a hash in a logstash configuration manually with at least two entries, and put a comment between them. It will work on the patched logstash, and give a syntax error on the unpatched logstash.

## Related issues

- Closes #15651 
- Closes #15561 

## Use cases

For example,

```
filter {
  mutate {
    add_field => {
      "field1" => "value1"
      # You can now put a comment between hash entries, like here
      "field2" => "value"
    }
  }
}
# You can now also put a comment before an EOF (i.e. no trailing newline)
```

### CODE REVIEW

File: logstash-core/lib/logstash/compiler/lscl/lscl_grammar.treetop [hunk 1]
```diff
@@ -7,7 +7,7 @@ grammar LogStashCompilerLSCLGrammar
   end
 
   rule comment
-    (whitespace? "#" [^\r\n]* "\r"? "\n")+ <LogStash::Compiler::LSCL::AST::Comment>
+    (whitespace? "#" [^\r\n]* "\r"? ("\n" / !.))+ <LogStash::Compiler::LSCL::AST::Comment>
```
Comment (andsel) [inline logstash-core/lib/logstash/compiler/lscl/lscl_grammar.treetop:10]: Could you explain better why this change is necessary? 
```
("\n" / !.)
```
Should mean "newline character **or**  anything that is not a character".

File: logstash-core/spec/logstash/config/config_ast_spec.rb [hunk 2]
```diff
@@ -211,6 +211,51 @@
...
+      def plugin(*args); end
+      it "hasn't got a trailing newline in the test config" do
+        # just to be clear that the last comment is supposed to end with EOF not \n
+        expect(config).not_to(end_with "\n")
+      end
+      it "should permit the comments" do
+        result = subject.parse(config)
+        expect(result).not_to(be_nil)
+        expect { eval(result.compile) }.not_to(raise_error)
+      end
```
Comment (andsel) [inline logstash-core/spec/logstash/config/config_ast_spec.rb:257]:
```suggestion
      let(:config) do %q(
        # early
        input # bla
        { # bla
          plugin # bla
          # bla
          { # bla
            # bla
            arg  # bla
            => # bla
            bareword # bla
            # bla
            hasharg => {
              #preentry
              "firstkey" => "firstvalue"
              # bla
              "middle" # bla
              # bla
              => # bla
              # bla
              "midvalue"  # bla
              # bla
              "lastkey" => "lastvalue"
              # bla
            }
          } # bla
        # bla
        } # bla
        # comment ending in EOF:
        ).strip 
      end
      
      subject { LogStashConfigParser.new }
      
      def line_to_source(*args); end   
      def plugin(*args); end
      
      it "hasn't got a trailing newline in the test config" do
        # just to be clear that the last comment is supposed to end with EOF not \n
        expect(config).not_to(end_with "\n")
      end
      
      it "should permit the comments" do
        result = subject.parse(config)
        expect(result).not_to(be_nil)
        expect { eval(result.compile) }.not_to(raise_error)
      end
```

**Review by andsel — COMMENTED** [review]
@jonaslb thank's a lot for this contribution.
Left a couple of comments.

Tried with the following:
```
input {
  stdin {
    codec => json {
      target => "[document]"
      # this comment already worked
      ecs_compatibility => disabled
    }
  }
}

filter {
  mutate {
    rename =>  {
      # this comment already worked
      "fieldA" => "fieldB"
      # this comment never worker
      "fieldC" => "fieldD"
      # this comment already worked
    }
  }
}

output {
	stdout {}
}

# this comment never worked
```

File: logstash-core/lib/logstash/compiler/lscl/lscl_grammar.treetop (hunk 1 above)
Comment (jonaslb) [reply logstash-core/lib/logstash/compiler/lscl/lscl_grammar.treetop:10]: Do you think this requires a comment in the grammar?

**Comment by jonaslb** [comment]
The reason

> @jonaslb thank's a lot for this contribution. Left a couple of comments.
> 
> Tried with the following:
> 
> ```
> input {
>   stdin {
>     codec => json {
>       target => "[document]"
>       # this comment already worked
>       ecs_compatibility => disabled
>     }
>   }
> }
> ```

The reason that comment already worked is because this isn't a Hash, but arguments for a plugin, which happens to look essentially the same but they're different things in the grammar (minor differences in allowed value types iirc).

**Comment by andsel** [comment]
> The reason that comment already worked is because this isn't a Hash, but arguments for a plugin, which happens to look essentially the same but they're different things in the grammar (minor differences in allowed value types iirc).

Yes you are right, is something different. 👍

File: logstash-core/lib/logstash/compiler/lscl/lscl_grammar.treetop (hunk 1 above)
Comment (andsel) [reply logstash-core/lib/logstash/compiler/lscl/lscl_grammar.treetop:10]: I thought was something along the EOF thing, so in this case you capture the comment at the end of file which terminates with EOF.

Would be better to put a comment on above the regexp to explain that `("\n" / !.)` is intended to capture the \n or the EOF in the comments that are the last in the file.

Comment (jonaslb) [reply logstash-core/lib/logstash/compiler/lscl/lscl_grammar.treetop]: changed this into a new rule and added a comment

**Comment by jonaslb** [comment]
Hi again @andsel , to be clear about what the pattern matches, I created a rule with a telling name for it (and used it in the comment rule):

```
  rule newline_or_eoi
    # `!.` is a negative lookahead for 'anything', i.e. it matches at the end of input.
    ("\r"? "\n") / !.
  end
```

What do you think of this?

**Comment by andsel** [comment]
BK run https://buildkite.com/elastic/logstash-pull-request-pipeline/builds/1061

**Comment by jonaslb** [comment]
Nothing seems to have been triggered by the "BK run" comment or the results are not visible here?

**Comment by jsvd** [comment]
buildkite test this

**Comment by andsel** [comment]
Thanks a lot @jonaslb for your contribution! It was accepted and merged 👏
